src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+        return MemoryType::Heap;
+    case Serialize::MemoryType::MemoryType_Stack:
+        return MemoryType::Stack;
+    case Serialize::MemoryType::MemoryType_Register:
+        return MemoryType::Register;
+    case Serialize::MemoryType::MemoryType_GPUShared:
+        return MemoryType::GPUShared;
+    case Serialize::MemoryType::MemoryType_GPUTexture:
+        return MemoryType::GPUTexture;
+    case Serialize::MemoryType::MemoryType_LockedCache:
+        return MemoryType::LockedCache;
+    case Serialize::MemoryType::MemoryType_VTCM:
+        return MemoryType::VTCM;
+    case Serialize::MemoryType::MemoryType_AMXTile:
+        return MemoryType::AMXTile;
+    default:
+        user_assert(false) << "unknown memory type " << memory_type << "\n";
+        return MemoryType::Auto;
+    }
+}
+
+ForType Deserializer::deserialize_for_type(const Serialize::ForType for_type) {
+    switch (for_type) {
+    case Serialize::ForType::ForType_Serial:
+        return ForType::Serial;
+    case Serialize::ForType::ForType_Parallel:
+        return ForType::Parallel;
+    case Serialize::ForType::ForType_Vectorized:
+        return ForType::Vectorized;
+    case Serialize::ForType::ForType_Unrolled:
+        return ForType::Unrolled;
+    case Serialize::ForType::ForType_Extern:
+        return ForType::Extern;
+    case Serialize::ForType::ForType_GPUBlock:
+        return ForType::GPUBlock;
+    case Serialize::ForType::ForType_GPUThread:
+        return ForType::GPUThread;
+    case Serialize::ForType::ForType_GPULane:
+        return ForType::GPULane;
+    default:
+        user_assert(false) << "unknown for type " << for_type << "\n";
+        return ForType::Serial;
+    }
+}
+
+DeviceAPI Deserializer::deserialize_device_api(const Serialize::DeviceAPI device_api) {
+    switch (device_api) {
+    case Serialize::DeviceAPI::DeviceAPI_None:
+        return DeviceAPI::None;
+    case Serialize::DeviceAPI::DeviceAPI_Host:
+        return DeviceAPI::Host;
+    case Serialize::DeviceAPI::DeviceAPI_Default_GPU:
+        return DeviceAPI::Default_GPU;
+    case Serialize::DeviceAPI::DeviceAPI_CUDA:
+        return DeviceAPI::CUDA;
+    case Serialize::DeviceAPI::DeviceAPI_OpenCL:
+        return DeviceAPI::OpenCL;
+    case Serialize::DeviceAPI::DeviceAPI_OpenGLCompute:
+        return DeviceAPI::OpenGLCompute;
+    case Serialize::DeviceAPI::DeviceAPI_Metal:
+        return DeviceAPI::Metal;
+    case Serialize::DeviceAPI::DeviceAPI_Hexagon:
+        return DeviceAPI::Hexagon;
+    case Serialize::DeviceAPI::DeviceAPI_HexagonDma:
+        return DeviceAPI::HexagonDma;
+    case Serialize::DeviceAPI::DeviceAPI_D3D12Compute:
+        return DeviceAPI::D3D12Compute;
+    case Serialize::DeviceAPI::DeviceAPI_Vulkan:
+        return DeviceAPI::Vulkan;
+    case Serialize::DeviceAPI::DeviceAPI_WebGPU:
+        return DeviceAPI::WebGPU;
+    default:
+        user_assert(false) << "unknown device api " << device_api << "\n";
```

Comment:
Style nit: here and elsewhere, just use `user_error` instead of `user_assert(false)`.